### ISSUE TITLE
mod_sftp does not handle multiple concurrent open file handles/transfers well for logging

### ISSUE THREAD

**Issue by alfonso-escribano** [proftpd/proftpd#1646]
### What I Did

Copy or sync with rclone using sftp several files to sftp user:
```
alfons@alfonso-escribano ~/test $ rclone -vvvv --verbose --sftp-disable-concurrent-writes --sftp-disable-concurrent-reads --sftp-use-fstat --sftp-set-modtime=false copy  . sftp-test:observation/
<7>DEBUG : rclone: Version "v1.62.2-DEV" starting with parameters ["rclone" "-vvvv" "--verbose" "--sftp-disable-concurrent-writes" "--sftp-disable-concurrent-reads" "--sftp-use-fstat" "--sftp-set-modtime=false" "copy" "." "sftp-test:observation/"]
<7>DEBUG : rclone: systemd logging support activated
<7>DEBUG : Creating backend with remote "."
<7>DEBUG : Using config file from "/home/alfons/.config/rclone/rclone.conf"
<7>DEBUG : fs cache: renaming cache item "." to be canonical "/home/alfons/test"
<7>DEBUG : Creating backend with remote "sftp-test:observation/"
<7>DEBUG : sftp-test: detected overridden config - adding "{k87wA}" suffix to name
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: New connection 172.16.16.10:35754->172.16.16.10:2121 to "SSH-2.0-mod_sftp"
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Shell type "none" from config
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Relative path resolved to "/var/ftp/clientes/t88863/observation"
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Using root directory "/var/ftp/clientes/t88863/observation"
<7>DEBUG : fs cache: renaming cache item "sftp-test:observation/" to be canonical "sftp-test{k87wA}:observation/"
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Waiting for checks to finish
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Waiting for transfers to finish
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: New connection 172.16.16.10:35768->172.16.16.10:2121 to "SSH-2.0-mod_sftp"
<6>INFO  : aaaaa1.txt: Copied (new)
<6>INFO  : aaaaa2.txt: Copied (new)
<6>INFO  : 
Transferred:           10 MiB / 10 MiB, 100%, 0 B/s, ETA -
Transferred:            2 / 2, 100%
Elapsed time:         0.4s

<7>DEBUG : 23 go routines active
<7>DEBUG : sftp://t88863@172.16.16.10:2121/observation/: Closing 2 unused connections
alfons@alfonso-escribano ~/test $ 
```

But in transfer log only one file is logged:
```
Tue Mar 28 09:25:03 2023 0 172.16.16.10 3833856 /var/ftp/clientes/t88863/observation/aaaaa2.txt b _ i r t88863 sftp 0 * c
```
Also, size logged isn't corect:
```alfonso-escribano ~ # ls -lart /var/ftp/clientes/t88863/observation/aaaaa2.txt
-rw-r--r-- 1 2001 2001 5242880 mar 28 09:25 /var/ftp/clientes/t88863/observation/aaaaa2.txt
```

But both files are upload without problems:
```
-rw-r--r-- 1 2001 2001 5242880 mar 28 09:25 aaaaa1.txt
-rw-r--r-- 1 2001 2001 5242880 mar 28 09:25 aaaaa2.txt
```

### What I Expected/Wanted

Both files should appear in logs and with correct sizes and transfer times.

### ProFTPD Version and Configuration
Version and configuration is irrelevant to see this bug, you only need to have sftp
I put 1.3.8 version:
```
alfonso-escribano ~ # proftpd -V
Compile-time Settings:
  Version: 1.3.8 (stable)
  Platform: LINUX [Linux 6.2.7-gentoo-dist x86_64]
  OS/Release:
    NAME=Gentoo
    ID=gentoo
    PRETTY_NAME="Gentoo Linux"
    VERSION_ID="2.13"
  Built: lun mar 27 2023 16:05:11 CEST
  Built With:
    configure  '--prefix=/usr' '--build=x86_64-pc-linux-gnu' '--host=x86_64-pc-linux-gnu' '--mandir=/usr/share/man' '--infodir=/usr/share/info' '--datadir=/usr/share' '--sysconfdir=/etc' '--localstatedir=/var/lib' '--datarootdir=/usr/share' '--disable-static' '--docdir=/usr/share/doc/proftpd-1.3.8' '--htmldir=/usr/share/doc/proftpd-1.3.8/html' '--libdir=/usr/lib64' '--localstatedir=/run/proftpd' '--sysconfdir=/etc/proftpd' '--disable-strip' '--enable-facl' '--enable-auth-file' '--enable-cap' '--disable-ctrls' '--enable-dso' '--disable-ident' '--enable-ipv6' '--disable-memcache' '--enable-ncurses' '--enable-nls' '--enable-openssl' '--enable-auth-pam' '--enable-pcre' '--disable-sodium' '--disable-tests' '--enable-trace' '--enable-shadow' '--enable-autoshadow' '--with-modules=mod_facl:mod_ifsession:mod_ldap:mod_sql:mod_sql_passwd:mod_sql_postgres:mod_sql_sqlite:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_snmp:mod_tls:mod_tls_shmcache' 'build_alias=x86_64-pc-linux-gnu' 'host_alias=x86_64-pc-linux-gnu' 'CC=x86_64-pc-linux-gnu-gcc' 'CFLAGS=-O2 -march=native ' 'LDFLAGS=-Wl,-O1 -Wl,--as-needed' 'CXXFLAGS=-O2 -march=native '

  CFLAGS: -g2 -O2 -march=native  -Wall -fno-omit-frame-pointer -fno-strict-aliasing
  LDFLAGS: -Wl,-L$(top_srcdir)/lib,-L$(top_builddir)/lib -Wl,-O1 -Wl,--as-needed -rdynamic  -L/usr/lib64/postgresql-15/lib64
  LIBS:  -lpcreposix -lpcre -lssl -lcrypto -lcap  -lssl  -lsqlite3  -lm -lpq  -lldap -llber  -lpam  -lcrypto -lattr -lcrypt  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd/proftpd.pid
    Scoreboard File:
      /run/proftpd/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/libexec

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    - Controls support
    - curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 3.0.8 7 Feb 2023)
    + PCRE support
    - PCRE2 support
    + POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    - Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

## Debug of the bug I have found.
Rclone use several process to connect to the server, sometimes upload each file with each process, but sometimes it uploads both files with the same process (I think it is also a bug in last version of rclone, I have to open an issue to rclone).
But also proftpd server have an strange behaviour that I think is a bug.
Using tace for command, and sftp I can see what is happening.
I put only relevant trace messages
First, process 57196 open a file to upload data:
```
2023-03-28 09:25:03,645 [57196] <sftp:6>: received OPEN (3) SFTP request (request ID 9, channel ID 0)
2023-03-28 09:25:03,645 [57196] <sftp:15>: OPEN flags = FXF_WRITE|FXF_CREAT|FXF_TRUNC
2023-03-28 09:25:03,645 [57196] <sftp:7>: received request: OPEN /var/ftp/clientes/t88863/observation/aaaaa1.txt type=unknown; (O_WRONLY|O_CREAT|O_TRUNC)
2023-03-28 09:25:03,645 [57196] <sftp:8>: sending response: HANDLE c287d0e808ec13b7
```
Before that file is completed, same process start other upload:
```
2023-03-28 09:25:03,782 [57196] <sftp:6>: received OPEN (3) SFTP request (request ID 112, channel ID 0)
2023-03-28 09:25:03,782 [57196] <sftp:15>: OPEN flags = FXF_WRITE|FXF_CREAT|FXF_TRUNC
2023-03-28 09:25:03,782 [57196] <sftp:7>: received request: OPEN /var/ftp/clientes/t88863/observation/aaaaa2.txt type=unknown; (O_WRONLY|O_CREAT|O_TRUNC)
2023-03-28 09:25:03,782 [57196] <sftp:8>: sending response: HANDLE 1ab0208ef5bb2583
```
Then, first file upload is completed:
```
2023-03-28 09:25:03,855 [57196] <sftp:6>: received CLOSE (4) SFTP request (request ID 233, channel ID 0)
2023-03-28 09:25:03,855 [57196] <sftp:7>: received request: CLOSE c287d0e808ec13b7
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_sql.c
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_sql.c
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_xfer.c
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_sql.c
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_log.c
2023-03-28 09:25:03,855 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_c3filtro.c
2023-03-28 09:25:03,864 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_snmp.c
2023-03-28 09:25:03,864 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa1.txt' to mod_xfer.c
```
And later, second file upload is completed:
```
2023-03-28 09:25:03,927 [57196] <sftp:6>: received CLOSE (4) SFTP request (request ID 338, channel ID 0)
2023-03-28 09:25:03,927 [57196] <sftp:7>: received request: CLOSE 1ab0208ef5bb2583
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_sql.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_sql.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching POST_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_xfer.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_sql.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_log.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_c3filtro.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_snmp.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'STOR /var/ftp/clientes/t88863/observation/aaaaa2.txt' to mod_xfer.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching POST_CMD command 'CLOSE 1ab0208ef5bb2583' to mod_sql.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'CLOSE 1ab0208ef5bb2583' to mod_sql.c
2023-03-28 09:25:03,927 [57196] <command:7>: dispatching LOG_CMD command 'CLOSE 1ab0208ef5bb2583' to mod_log.c
```

What I think is happening:
When sftp client opens two or more filehandlres in same session, upload are done correctly due to filehandlers are manage by sftp module, but in mod_xfer and other modules, `session.xfer.path` it's only one, that table doesn't have support to upload several files at 
 same time

I have seen that bug, due to we have our own proftpd module that when file is uploaded, it creates a file with data like uploaded file, user and other info, and we have seen that using rclone if fails.

We are obtaining uploaded file in our own module, from:
```
fullname = dir_abs_path(session.xfer.p,session.xfer.path,TRUE);
```
But in that cases, it is null.
Changing it to:
```
path = pr_table_get(cmd->notes, "mod_xfer.store-path", NULL);
```
We can have correct file.

Also for log, we use custom `ExtendedLog` to have our own transferlog, and if we use "%b" for file size, it is incorrect, `%{transfer-millisecs}` is also incorrect and `%F` is empty.

I have seen that this bug is present in 1.3.6e and 1.3.8, I haven't test with 1.3.7 but I suppose that the problem is the same.

In mod_sftp/fxp.c I have seen a reference to Bug#3707 and I think it's related to the bug I have found:
```
  /* Add a note containing the file handle for logging (Bug#3707). */
  fxp_set_filehandle_note(cmd, fxh);
```
> Note that I also fixed the logging of %f LogFormat variables for SFTP CLOSE requests; the session.xfer struct was being cleared by the RETR/STOR handles to which the CLOSE request handler dispatched and thus when the CLOSE request was dispatched, mod_log had no session.xfer information to work with.  Now fixed.

Maybe you need to fix how `session.xfer` is used when you have several filehandles in same session, or use more notes to keep track of diferrent filehandlers in same session.

If you need any other info or more specific tests, let me know.

It's possible that this bug is affecting a lot of people in a silent way, we have seen it due to our custom proftpd module.

Also it can have minor security considerations due to it's possible to use this bug to upload files without letting info in any log file in the server.

**Comment by Castaglia** [comment]
It's true that for logging purposes, `mod_sftp` does not currently handle the case of multiple open filehandles gracefully.  That is, the files will be transferred successfully via SFTP but, as you point out, the _assumptions_ made by the use of a single global `session.xfer` struct, which work for FTP, do not work well for SFTP.  In FTP, data transfers are completed handled by a single command whereas with SFTP, data transfers are implemented via multiple protocol requests (`OPEN`, `READ`/`WRITE`s, `CLOSE`).

I'll see what I can do to improve the SFTP implementation for these situations.

**Comment by Castaglia** [comment]
@alfonso-escribano in my local reproduction setup, https://github.com/proftpd/proftpd/pull/1652 looks to address this behavior.  Would you like to test it out in your environment, see if it fixes things for you?

**Comment by alfonso-escribano** [comment]
@Castaglia I have test the patch, but it isn't correct for the case you are using HiddenStores, in that case, finally you end with the path of real hidden file in transferlog and other modules in "session.xfer.path" instead of the corresponding file that you have in mod_xfer.store-path or in cmd.
Whith HiddenStores
`HiddenStores            ".temp-filepart-%P."`

and the patch, I get this on transferlog:
**_Mon Apr 03 09:27:08 2023 0 172.16.16.10 5242880 /var/ftp/clientes/t88863/observation/.temp-filepart-167066.aaaaa2.txt. b _ i r t88863 sftp 0 * c
Mon Apr 03 09:27:08 2023 0 172.16.16.10 5242880 /var/ftp/clientes/t88863/observation/.temp-filepart-167066.aaaaa1.txt. b _ i r t88863 sftp 0 * c_**

Whit the patch in our custom proftpd module, I have this values:
**mod_xfer.store-path** => "/observation/aaaaa1.txt"
**session.xfer.path** => "/observation/.temp-filepart-167066.aaaaa1.txt."
**pr_cmd_get_displayable_str(cmd, NULL)** => "STOR /observation/aaaaa1.txt"
I think this is the one issue that remains, now size of the file and transfer times are corrected

**Comment by Castaglia** [comment]
@alfonso-escribano I've just updated the PR with what should a fix/correction, such that `HiddenStores` are logged properly, if you'd like to test again?

**Comment by alfonso-escribano** [comment]
@Castaglia I have check it, and now it logs properly and our custom module gets the correct info.
You can merge and backport this PR when you consider it appropiate.
Thanks for your work.

**Comment by Castaglia** [comment]
This has been merged to master, and backported to the 1.3.8 branch.  Thanks!
